File: CANOAS MALUCAS/JAVA/CanoasMalucas3_0/src/Rio.java
public class Rio {
    public boolean[] blocos;
    public Canoa canoa;
    private int ultimaPosicaoMovimentoCanoa;
    public Rio(int tamanho, Canoa canoa){
        if (tamanho > (canoa.blocos.length * 2)) {
            blocos = new boolean[tamanho];
            this.canoa = canoa;
            moverCanoa(0);
        }        
    }
    public int getUltimaPosicaoMovimentoCanoa(){
        return ultimaPosicaoMovimentoCanoa;
    }
    public void limparRio(){
        for (int indice = 0; indice < blocos.length; indice++) {
            blocos[indice] = false;
        }
    }
    public void moverCanoa(int posicao){        
        int limiteCanoa = posicao + canoa.blocos.length - 1;
        if (posicao >= 0 && limiteCanoa < blocos.length) {
            ultimaPosicaoMovimentoCanoa = posicao;
            int indiceCanoa = 0;            
            for (int indiceRio = 0; indiceRio < blocos.length; indiceRio++) {
                if (indiceRio >= posicao && indiceCanoa < canoa.blocos.length) {
                    blocos[indiceRio] = canoa.blocos[indiceCanoa];
                    indiceCanoa++;
                }
                else{
                    blocos[indiceRio] = false;
                }
            }
        }
    }
    public boolean estaVazio(){
        for (boolean bloco : blocos) {
            if (bloco == true) {
                return false;
            }
        }        
        return true;
    }
    /*
                [0][1][2][3]
    [~][~][~][~][#][#][~][#][~][~]
    [0][1][2][3][4][5][6][7][8][9]
                 U     A  F
    */
    public void receberAtaque(int posicao){
        if (posicao >=0 && posicao < blocos.length) {
            int posicaoFinalCanoa = ultimaPosicaoMovimentoCanoa + canoa.blocos.length;
            if (posicao >= ultimaPosicaoMovimentoCanoa && posicao < posicaoFinalCanoa) {
                int posicaoDentroCanoa = posicao - ultimaPosicaoMovimentoCanoa;
                canoa.destruirBloco(posicaoDentroCanoa);                
            }
            moverCanoa(ultimaPosicaoMovimentoCanoa);
        }
    }
    @Override
    public String toString() {
        String textCanoa = "";
        for (int indice = 0; indice < canoa.blocos.length; indice++) {
            textCanoa += indice + ":[" + canoa.blocos[indice] + "]  ";
        }  
        String textBlocos = "";
        for (int indice = 0; indice < blocos.length; indice++) {
            textBlocos += indice + ":[" + blocos[indice] + "]  ";
        }    
        return  "Rio{\n" + 
                "  + canoa (" + canoa.nome + ") = " + textCanoa + "\n" +
                "  + blocos = " + textBlocos + "\n" +
                "  + Esta Vazio? " + estaVazio() + "\n" +
                "}";
    }
}

File: CANOAS MALUCAS/C#/CanoasMalucas/CanoasMalucas/Rio.cs
public class Rio
{
    public bool[] blocos;
    public Canoa canoa;
    private int ultimaPosicaoMovimentoCanoa;
    public Rio(int tamanho, Canoa canoa)
    {
        if (tamanho > (canoa.blocos.Length * 2))
        {
            blocos = new bool[tamanho];
            this.canoa = canoa;
            MoverCanoa(0);
        }
    }
    public int GetUltimaPosicaoMovimentoCanoa()
    {
        return ultimaPosicaoMovimentoCanoa;
    }
    public void LimparRio()
    {
        for (int indice = 0; indice < blocos.Length; indice++)
        {
            blocos[indice] = false;
        }
    }
    public void MoverCanoa(int posicao)
    {
        int limiteCanoa = posicao + canoa.blocos.Length - 1;
        if (posicao >= 0 && limiteCanoa < blocos.Length)
        {
            ultimaPosicaoMovimentoCanoa = posicao;
            int indiceCanoa = 0;
            for (int indiceRio = 0; indiceRio < blocos.Length; indiceRio++)
            {
                if (indiceRio >= posicao && indiceCanoa < canoa.blocos.Length)
                {
                    blocos[indiceRio] = canoa.blocos[indiceCanoa];
                    indiceCanoa++;
                }
                else
                {
                    blocos[indiceRio] = false;
                }
            }
        }
    }
    public bool EstaVazio()
    {
        foreach (bool bloco in blocos)
        {
            if (bloco == true)
            {
                return false;
            }
        }
        return true;
    }
    /*
                [0][1][2][3]
    [~][~][~][~][#][#][~][#][~][~]
    [0][1][2][3][4][5][6][7][8][9]
                 U     A  F
    */
    public void ReceberAtaque(int posicao)
    {
        if (posicao >= 0 && posicao < blocos.Length)
        {
            int posicaoFinalCanoa = ultimaPosicaoMovimentoCanoa + canoa.blocos.Length;
            if (posicao >= ultimaPosicaoMovimentoCanoa && posicao < posicaoFinalCanoa)
            {
                int posicaoDentroCanoa = posicao - ultimaPosicaoMovimentoCanoa;
                canoa.DestruirBloco(posicaoDentroCanoa);
            }
            MoverCanoa(ultimaPosicaoMovimentoCanoa);
        }
    }

    public override string ToString()
    {
        string textCanoa = "";
        for (int indice = 0; indice < canoa.blocos.Length; indice++)
        {
            textCanoa += indice + ":[" + canoa.blocos[indice] + "]  ";
        }
        string textBlocos = "";
        for (int indice = 0; indice < blocos.Length; indice++)
        {
            textBlocos += indice + ":[" + blocos[indice] + "]  ";
        }
        return "Rio{\n" +
                "  + canoa (" + canoa.nome + ") = " + textCanoa + "\n" +
                "  + blocos = " + textBlocos + "\n" +
                "  + Esta Vazio? " + EstaVazio() + "\n" +
                "}";
    }
}